Precompute vector length on indexing

mayya-sharipova opened the pull request:
Some vector functions need documents' vector length in their calculations,
and every time we run a query with those functions, we need to compute
these lengths again.  For high dims vectors it is an expensive calculation.

Instead the goal of this PR is precompute documents' vectors' lengths
in advance only once and store them together with storing vectors.

This change is not breaking as the procedure of encoding/decoding
vectors changes internally depending on the created version of index.
No user action is needed.

mayya-sharipova committed:
Precomput vector length on indexing

elasticmachine commented:
Pinging @elastic/es-search

mayya-sharipova commented:
Some evaluations:

**glove-100-angular**

Evaluation metric | es-bruteforce 5 shards | es-bruteforce with vector lengths precomputed 5 shards
-- | --: | --:
index size merged to 1 segm | 2.1G | 2.1Gb
queries per second | 6.66 | 8.3
index build time | 363.87 secs | 361.6 secs


**glove-200-angular**

Evaluation metric | es-bruteforce 5 shards | es-bruteforce with vector lengths precomputed 5 shards
-- | --: | --:
index size merged to 1 segm | 4.7gb | 4.2Gb
queries per second | 3.56 | 4.67
index build time | 517.75s | 523.15s

dakrone commented:
@mayya-sharipova can you fill in more information about this PR in the description? It's difficult to tell from only the title what exactly this does and how it's a breaking change from the perspective of someone not on the search team.

jimczi reviewed (commented):
I looked briefly but I don't think this needs to be a breaking change. You could check the created index version to switch from one format to another without requiring early adopters to reindex their data ? I also wonder if this should be an option in the field definition since it is not useful for every functions and adds an overhead in terms of index size. Can you also run your experiment with a single shard on an index that was forced merged to 1 segments ? The result you have seem to show that the new format takes less space than the new one, this doesn't make sense since the format adds 4 bytes to any vectors so we should eliminate merging diff with a force merge.

mayya-sharipova commented:
@elasticmachine run elasticsearch-ci/1

mayya-sharipova commented:
@elasticmachine run  elasticsearch-ci/2

mayya-sharipova commented:
@jimczi Thanks for the feedback, addressing some of your comments.

>  You could check the created index version to switch from one format to another without requiring early adopters to reindex their data ?

Currently decoding of vector's BytesRef is done inside corresponding painless functions. And these painless functions are static/stateless towards index parameters.  We can change these painless functions to accept  indexVersion as a hidden parameter, but I am wondering if it is worth the effort, as the vectors functions are still experimental and are allowed to be changed in non-compatible way.

> I also wonder if this should be an option in the field definition since it is not useful for every functions and adds an overhead in terms of index size

I am worried about introducing an extra parameter that users should think about.  Also, ann-bechmarks when the score is not necessary but only the order matters, [optimizes](https://github.com/erikbern/ann-benchmarks/blob/master/ann_benchmarks/algorithms/bruteforce.py#L63-L67)  other functions to use vectors' lengths as well.

mayya-sharipova commented:
> Can you also run your experiment with a single shard on an index that was forced merged to 1 segments ?

@jimczi here are the results for this experiment.


**glove-200-angular**

Evaluation metric | es-bruteforce | es-bruteforce with vector lengths precomputed 
-- | --: | --:
index size merged to 1 segm | 4.2Gb | 4.2Gb
queries per second | 1.498 | :1.083
index build time | 492s | 488s

jimczi commented:
> We can change these painless functions to accept indexVersion as a hidden parameter, but I am wondering if it is worth the effort, as the vectors functions are still experimental and are allowed to be changed in non-compatible way.

I think we should try to remain compatible if we can and this doesn't look like a big change. Keeping the index created version when we apply vector functions will allow us to do more changes in the future in a backward compatible way so I'd be +1 to try this route.
The performance improvement looks good, I wonder though if this is needed for all vectors since the difference should be minimal when the number of dimensions is small. This is also equivalent to adding a single dimension to all vectors so that's maybe an acceptable trade-off.

mayya-sharipova committed:
Add scoreScript parameter to vector functions

This parameter is hidden and doesn't need to be supplied by users.
It allows to check index version and use different encodings/
decodings depending on the version.

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into precompute-vector-length-on-indexing

mayya-sharipova commented:
@jimczi  Thanks for the feedback, Jim. I have added `indexVersion` to vector functions, and depending on the decoding happens differently.

The only thing left unaddressed is if we want to make vector lengths' pre-computation and storage optional. I am ok to make optional as defined by a user in the mapping:
```
     "my_vector": {
        "type": "dense_vector",
        "dims": 100,
        "store_length" : true    # false by default
      }
```
WDYT?

jtibshirani commented:
@mayya-sharipova to me it seems best to always include the vector length instead of introducing a mapping option. We are mainly targeting high-dimensional vectors with this type, so indexing one additional value seems acceptable -- as @jimczi noted, this would be equivalent to having a vector of length 501 as opposed to 500. We usually try to cut down on different options unless it is clear they are valuable in practice.

jimczi commented:
I agree with @jtibshirani , let's always include the vector length for indices created on or after 7.4.0.

mayya-sharipova commented:
Thanks @jimczi and @jtibshirani.  If we always include length from 7.4, that this PR is ready for review. Can you please continue to review it when you have time.

mayya-sharipova committed:
Allow to add vectors to indexes pre 7_4

mayya-sharipova commented:
@jimczi @jtibshirani  This PR is ready for another set of review.

mayya-sharipova committed:
Fix check style

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into precompute-vector-length-on-indexing

jtibshirani reviewed (commented):
Seems like a nice speed improvement! I left a couple suggestions around unifying the code paths.

jtibshirani commented:
One other comment that came to mind -- I wonder if we could use the precomputed vector norms to help speed up `l2norm` as well? The L2 distance can be written as

```
sqrt(query.query - 2 query.doc + doc.doc)
```

where `.` represents dot product. So we could precompute `query . query` once, and use the precomputed vector norms to quickly calculate `doc.doc`. I don't think it will help as much as in the cosine case, since we still need to compute a square root per document, but it could be worth trying.

mayya-sharipova committed:
Remove duplicate code

mayya-sharipova commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/mapper/DenseVectorFieldMapper.java:
```diff
@@ -180,21 +181,51 @@ public void parse(ParseContext context) throws IOException {
-            if (dim++ >= dims) {
-                throw new IllegalArgumentException("Field [" + name() + "] of type [" + typeName() + "] of doc [" +
-                    context.sourceToParse().id() + "] has exceeded the number of dimensions [" + dims + "] defined in mapping");
+        if (indexCreatedVersion.onOrAfter(Version.V_7_4_0)) {
```
Thanks @jtibshirani. Makes sense. Addressed with[baddf608](https://github.com/elastic/elasticsearch/pull/45390/commits/38f1e32577b5b9b9a31c15d8ceba54bc5c545d54)

mayya-sharipova commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/mapper/VectorEncoderDecoder.java:
```diff
@@ -20,13 +21,22 @@ private VectorEncoderDecoder() { }
      * @return BytesRef
      */
-    public static BytesRef encodeSparseVector(int[] dims, float[] values, int dimCount) {
+    public static BytesRef encodeSparseVector(Version indexVersion, int[] dims, float[] values, int dimCount) {
```
Thanks @jtibshirani. Makes sense. Addressed with[baddf608](https://github.com/elastic/elasticsearch/pull/45390/commits/38f1e32577b5b9b9a31c15d8ceba54bc5c545d54)

mayya-sharipova commented:
>  I wonder if we could use the precomputed vector norms to help speed up l2norm as well? 

@jtibshirani  Should we explore this in a subsequent PR? I can run benchmarks to see how what will be the speed improvements?

jtibshirani commented on x-pack/plugin/vectors/src/test/java/org/elasticsearch/xpack/vectors/mapper/VectorEncoderDecoderTests.java:
```diff
@@ -16,15 +17,39 @@
-        float[] decodedValues = VectorEncoderDecoder.decodeDenseVector(encodedDenseVector);
+        float[] decodedValues = VectorEncoderDecoder.decodeDenseVector(indexVersion, encodedDenseVector);
+        float decodedMagnitude = VectorEncoderDecoder.getVectorMagnitude(indexVersion, encodedDenseVector, decodedValues);
+        assertEquals(expectedMagnitude, decodedMagnitude, 0.001f);
```
Should this error threshold be `0.0f`, since we don't expect any loss of precision from encoding and decoding?

jtibshirani commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/mapper/VectorEncoderDecoder.java:
```diff
@@ -20,50 +21,76 @@ private VectorEncoderDecoder() { }
-        int offset = 0;
+        byte[] buf;
+        int offset;
+        if (indexVersion.onOrAfter(Version.V_7_4_0)) {
```
Looking over the code again, I am wondering if it'd be nicer to encode the vector norm at the end of the array, instead of the beginning. That way we wouldn't have to adjust the `offset` parameter when we're reading + writing the arrays.

mayya-sharipova replied:
@jtibshirani This is a good idea, and indeed simplified the code. Addressed in [a69ea98](https://github.com/elastic/elasticsearch/pull/45390/commits/a69ea98478bad7cf5a44972ab758dfea4e058b87)

jtibshirani reviewed (commented):
I have a couple last small comments, thanks for all the changes so far. 

> Should we explore this in a subsequent PR?

Sure, that sounds good. We had an offline discussion around opening a new short-lived feature branch like `optimize-vector-functions`, and rebasing this PR against it. So I will look into opening a PR against that branch as a follow-up.

mayya-sharipova committed:
Encode vector magnitude at the end
